pkg/status/status.go

```diff
@@ -0,0 +1,264 @@
+package status
+
+import (
+	"fmt"
+	"github.com/kyma-incubator/octopus/pkg/apis/testing/v1alpha1"
+	"github.com/kyma-incubator/octopus/pkg/consts"
+	"k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"time"
+)
+
+type NowProvider func() time.Time
+
+type Service struct {
+	nowProvider NowProvider
+}
+
+func NewService(nowProvider NowProvider) *Service {
+	return &Service{nowProvider: nowProvider}
+}
+
+func (s *Service) EnsureStatusIsUpToDate(suite v1alpha1.ClusterTestSuite, pods []v1.Pod) (*v1alpha1.TestSuiteStatus, error) {
+	out := suite.Status.DeepCopy()
+	for _, pod := range pods {
+		for resultID, tr := range out.Results {
+			if tr.Name == pod.Labels[consts.LabelKeyTestDefName] && tr.Namespace == pod.Namespace {
+				// find execution
+				for execID, exec := range tr.Executions {
+					if exec.ID == pod.Name {
+						prev := exec.PodPhase
+						if pod.Status.Phase != prev {
+							out.Results[resultID].Executions[execID] = s.adjustTestExec(exec, pod)
+						}
+					}
+				}
+			}
+		}
+	}
+
+	for idx, res := range out.Results {
+		newState := s.calculateTestStatus(res)
+		if res.Status != newState {
+			out.Results[idx].Status = newState
+		}
+	}
+	adjusted := s.adjustSuiteCondition(*out)
+	out = &adjusted
+	// TODO later what to do with incosistencies?
+	return out, nil
+}
+
+func (s *Service) adjustTestExec(exec v1alpha1.TestExecution, pod v1.Pod) v1alpha1.TestExecution {
+	exec.PodPhase = pod.Status.Phase
+	if exec.PodPhase == v1.PodSucceeded {
+		exec.CompletionTime = &metav1.Time{Time: s.nowProvider()}
+	} else if exec.PodPhase == v1.PodFailed {
+		exec.CompletionTime = &metav1.Time{Time: s.nowProvider()}
+		exec.Reason = pod.Status.Reason
+		exec.Message = pod.Status.Message
+	}
+	return exec
+}
+
+func (s *Service) calculateTestStatus(tr v1alpha1.TestResult) v1alpha1.TestStatus {
```

Comment:
I do not get sth or it can be simple as that?
```go
func (s *Service) calculateTestStatus(tr v1alpha1.TestResult) v1alpha1.TestStatus {
	if len(tr.Executions) == 0 {
		return v1alpha1.TestNotYetScheduled
	}

	for _, exec := range tr.Executions {
		switch exec.PodPhase {
		case v1.PodSucceeded:
			// success, check if other also will have this phase
			continue
		case v1.PodPending:
			return v1alpha1.TestRunning
		case v1.PodFailed:
			return v1alpha1.TestFailed
		case v1.PodRunning:
			return v1alpha1.TestRunning
		case v1.PodUnknown:
			return v1alpha1.TestUnknown
		default:
			return v1alpha1.TestUnknown
		}
	}

	return v1alpha1.TestSucceeded
}
```
---

NOTE:
The statement
```go
	if anySucceeded {
		return v1alpha1.TestSucceeded
	}
```
is confusing cause it means that even if only 1/12 was succeeded that the whole test can be also marked as successful and this is not true, should be more like

```go
	if allSucceeded {
		return v1alpha1.TestSucceeded
	}
```
and only because u have the correct order of `if` statements it has a correct behaviour but it can be easly broken after some refactore phase